### ISSUE TITLE
[Feature request] Kwikset Convert Lock

### ISSUE THREAD

**Issue by Elder-HVAC-Man** [itavero/homebridge-z2m#9]
**Problem description**
Kwikset convert pairs but can not be controlled

**Suggested solution**
Would like for the lock to work properly

**Partial Log:**
```
{"ieeeAddr":"0x002446fffd033c48","type":"EndDevice","networkAddress":64587,"model":"66492-001","vendor":"Kwikset","description":"Home connect smart lock conversion kit","friendly_name":"XXX","manufacturerID":4242,"manufacturerName":"Kwikset","powerSource":"Battery","modelID":"SMARTCODE_CONVERT_GEN1","hardwareVersion":3,"lastSeen":1594565932326}]}
```

```
{"state":"LOCK","lock_state":"locked","linkquality":178}
```

```
error 2020-07-12 10:06:47: Publish 'set' 'state' to 'XXX' failed: 'Error: Cluster 'closuresDoorLock' has no command 'onDoor'

error 2020-07-12 10:06:52: Publish 'set' 'state' to 'XXX' failed: 'Error: Cluster 'closuresDoorLock' has no command 'offDoor'
```

**Homebridge Log:**
```
[7/12/2020, 10:28:56 AM] [zigbee2mqtt] Publish to 'zigbee2mqtt/XXX/set': '{"state":"ON"}'
[7/12/2020, 10:29:10 AM] [zigbee2mqtt] Publish to 'zigbee2mqtt/XXX/set': '{"state":"OFF"}'
```

**Zigbee devices.js**
```
// Kwikset
    {
        zigbeeModel: ['SMARTCODE_CONVERT_GEN1'],
        model: '66492-001',
        vendor: 'Kwikset',
        description: 'Home connect smart lock conversion kit',
        supports: 'lock/unlock, battery',
        fromZigbee: [fz.lock, fz.lock_operation_event, fz.battery],
        toZigbee: [tz.generic_lock],
        meta: {configureKey: 3},
        configure: async (device, coordinatorEndpoint) => {
            const endpoint = device.getEndpoint(2);
            await bind(endpoint, coordinatorEndpoint, ['closuresDoorLock', 'genPowerCfg']);
            await configureReporting.lockState(endpoint);
            await configureReporting.batteryPercentageRemaining(endpoint);
        },
```

**Are you willing to contribute to this feature?**
Yes, but not sure what I can contribute as I am fairly new to this, especially Zigbee.

**Comment by itavero** [comment]
I don't have a lock myself, but I actually already thought that this might lead to some conflicts.

I guess depending on the value of the `state` attribute, it should either be mapped to a Switch or Lightbulb, which it does now, or a [Lock Mechanism](https://developers.homebridge.io/#/service/LockMechanism).

I might have some time next weekend to look into this. If I need some more information, I'll let you know.

**Comment by itavero** [comment]
Cleaned up your message a bit (added some formatting).

Can you provide me with the following:
* the report message of when the device is locked and from when the device is unlocked, so I can determine which values I can expect in `state` for a lock (and maybe also in `lock_state`).
* the message you need to send to MQTT in order to lock or unlock the device.

**Comment by Elder-HVAC-Man** [comment]
Thank you for looking into this.  I have spent the last few hours installing the lock and trying to operate it via Home Assistant and Homekit.  It was a very frustrating experience.  Lock wasn't operating smoothly, as it was binding for some unknown reason.  When I operated the lock away from the door, it operated wrong.  On an open command, the lock would rotate, say in a CW rotation, and show closed.  When asked to close, the lock would rotate in a CW rotation AGAIN.  Needless to say, after discovering this type of operation, I removed the lock and put the original lock back in place.

I don't have any app available for it to make it work, as it is supposed to be paired with the Amazon cloud cam or something like that.  I will look and see if I can find where the Home Assistant logs are kept, as they might have what you need.  I have the YAML for HA somewhere that supposedly works, if that will help.

Could the lock be looking for an end (binding) for open or close to force it in the other direction on the next command?

Thanks again.

**Comment by itavero** [comment]
I'm not sure I understand your last question. Can you rephrase it?

For this plugin to work properly, it would have to work with zigbee2mqtt and I would have to know what the MQTT communication looks like. If I understand it correctly, you haven't gotten the lock to work at all?

**Comment by itavero** [comment]
Is [this the device](https://www.zigbee2mqtt.io/devices/66492-001.html) you are talking about?

For me there's actually enough information on that page to probably get it to work.
As mentioned, I hope I can have look at it next weekend.

**Comment by Elder-HVAC-Man** [comment]
Yes, that is the lock.  Yes, the lock is working with Zigbee2mqtt.  When I had it installed on the door, the latch wasn’t smooth enough, so I messed with it for hours trying different hardware available.  I will try again once you have added it.  Thanks.

**Comment by itavero** [comment]
@Elder-HVAC-Man I made a "theoretical" implementation, as I do not have any Zigbee lock to verify it with.
It's available on the `master` branch. Are you able to verify if this works for you?

The README file contains information on [how to test the latest changes from the `master` branch](https://github.com/itavero/homebridge-z2m#testing-the-latest-changes).

**Comment by Elder-HVAC-Man** [comment]
I am not sure if I installed it correctly or not, but the log file shows this error:
Error: The requested platform 'zigbee2mqtt' was not registered by any plugin.
    at PluginManager.getPluginForPlatform (/usr/local/lib/node_modules/homebridge/src/pluginManager.ts:224:15)
    at /usr/local/lib/node_modules/homebridge/src/server.ts:387:37
    at Array.forEach (<anonymous>)
    at Server.loadPlatforms (/usr/local/lib/node_modules/homebridge/src/server.ts:374:27)
    at Server.start (/usr/local/lib/node_modules/homebridge/src/server.ts:153:29)
    at cli (/usr/local/lib/node_modules/homebridge/src/cli.ts:80:10)
    at Object.<anonymous> (/usr/local/lib/node_modules/homebridge/bin/homebridge:17:22)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
I installed it in homebridge terminal using this command: sudo npm i -g itavero/homebridge-z2m#master
By the way, I ended up getting it to work via Home Assistant, but I would much rather have only 1 bridge for my devices that are not native Homekit. 

Thanks,
Larry Novak 

    On Monday, July 13, 2020, 2:50:39 PM CDT, Arno Moonen <notifications@github.com> wrote:  
 
 


@Elder-HVAC-Man I made a "theoretical" implementation, as I do not have any Zigbee lock to verify it with.
It's available on the master branch. Are you able to verify if this works for you?

The README file contains information on how to test the latest changes from the master branch.

**Comment by itavero** [comment]
That's interesting. By the way, can you please use Markdown formatting in your comments (for instance, use a code block for the logs)? Makes it a bit easier to read.

I always use the `npm link` command, but I'm pretty sure I've tested the command you mentioned successfully as well.
May I ask how you are running Homebridge and which version you are using?

**Comment by Elder-HVAC-Man** [comment]
I am very new to all of this, so I don't know how to properly format text, but I see some icons above that will be helpful next time.

Homebridge version - 1.1.1, running on a Raspberry Pi 4b

**Comment by itavero** [comment]
Just noticed that Markdown formatting is not possible because you replied by email. I've never used that reply feature, so I wasn't aware.
I'll try to double check the command (probably this weekend).

**Comment by itavero** [comment]
Unfortunately I didn't come around to verifying the command / the plugin on a clean installation. As soon as I have, I'll let you know.

**Comment by Elder-HVAC-Man** [comment]
I appreciate you trying.  I am moving all of my devices from Homebridge to Home Assistant anyway.  The lock is supported there.
Thanks again.

**Comment by itavero** [comment]
Personally I found configuring Home Assistant correctly for each device (even with the MQTT Discovery a lot of hassle). But I can understand that you might prefer that route at this point. 😅 

I'll try to verify it myself based on the messages you've shared and the code from Zigbee2mqtt. At least I can keep the changes in, so it's there if others want to give it a go.

**Comment by itavero** [comment]
I manually published the following status messages to MQTT and they were picked up correctly by this Homebridge plugin (using what is currently on the `master` branch):

* `{"state":"UNLOCK","lock_state":"not_fully_locked","linkquality":178}`
* `{"state":"UNLOCK","lock_state":"unlocked","linkquality":178}`
* `{"state":"LOCK","lock_state":"locked","linkquality":178}`

Toggling the state sends a message to the `set` topic with either:
* `{"state":"LOCK"}`
* `{"state":"UNLOCK"}`

Closing this issue, because (as far as I can test it) the lock mechanism works.

**Comment by itavero** [comment]
Included in v0.0.6
